File: src/code-samples/langchain/long-term-memory-storage-inmemory.ts
// :snippet-start: long-term-memory-storage-inmemory-js
import { InMemoryStore } from "@langchain/langgraph";

const embed = (texts: string[]): number[][] => {
  // Replace with an actual embedding function or LangChain embeddings object
  return texts.map(() => [1.0, 2.0]);
};

// InMemoryStore saves data to an in-memory dictionary. Use a DB-backed store in production use.
const store = new InMemoryStore({ index: { embed, dims: 2 } });
const userId = "my-user";
const applicationContext = "chitchat";
const namespace = [userId, applicationContext];

await store.put(namespace, "a-memory", {
  rules: [
    "User likes short, direct language",
    "User only speaks English & TypeScript",
  ],
  "my-key": "my-value",
});

// get the "memory" by ID
const item = await store.get(namespace, "a-memory");

// search for "memories" within this namespace, filtering on content equivalence, sorted by vector similarity
const items = await store.search(namespace, {
  filter: { "my-key": "my-value" },
  query: "language preferences",
});
// :snippet-end:

// :remove-start:
async function main() {
  // Verify the operations work
  if (!item) {
    throw new Error("Item should not be null");
  }
  if (item.value["my-key"] !== "my-value") {
    throw new Error('Expected my-key to be "my-value"');
  }
  if (!item.value["rules"]) {
    throw new Error("Expected rules to exist");
  }
  if ((item.value["rules"] as string[]).length !== 2) {
    throw new Error("Expected 2 rules");
  }

  // Verify search returns results
  if (items.length === 0) {
    throw new Error("Expected search to return results");
  }
  if (items[0].value["my-key"] !== "my-value") {
    throw new Error('Expected search result my-key to be "my-value"');
  }

  console.log("✓ InMemoryStore operations work correctly");
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((error) => {
    console.error(error);
    process.exit(1);
  });
}
// :remove-end:

File: src/code-samples/langchain/long-term-memory-storage-inmemory.py
# :snippet-start: long-term-memory-storage-inmemory-py
from collections.abc import Sequence

from langgraph.store.base import IndexConfig
from langgraph.store.memory import InMemoryStore


def embed(texts: Sequence[str]) -> list[list[float]]:
    # Replace with an actual embedding function or LangChain embeddings object
    return [[1.0, 2.0] * len(texts)]


# InMemoryStore saves data to an in-memory dictionary. Use a DB-backed store in production use.
store = InMemoryStore(index=IndexConfig(embed=embed, dims=2))
user_id = "my-user"
application_context = "chitchat"
namespace = (user_id, application_context)
store.put(
    namespace,
    "a-memory",
    {
        "rules": [
            "User likes short, direct language",
            "User only speaks English & python",
        ],
        "my-key": "my-value",
    },
)
# get the "memory" by ID
item = store.get(namespace, "a-memory")
# search for "memories" within this namespace, filtering on content equivalence, sorted by vector similarity
items = store.search(
    namespace, filter={"my-key": "my-value"}, query="language preferences"
)
# :snippet-end:

# :remove-start:
if __name__ == "__main__":
    # Verify the operations work
    assert item is not None
    assert item.value["my-key"] == "my-value"
    assert "rules" in item.value
    assert len(item.value["rules"]) == 2

    # Verify search returns results
    assert len(items) > 0
    assert items[0].value["my-key"] == "my-value"

    print("✓ InMemoryStore operations work correctly")
# :remove-end:
